Dispatcher test cache failures.

@ghatwala reports the following in https://github.com/numba/numba/issues/2451#issuecomment-316605512, tracking in this ticket as the issue is separate to that reported #2451.

```
$ /home/amit/numba# ls -l /usr/local/lib/python2.7/dist-packages/ | grep llvmlite
drwxr-sr-x  6 root staff   4096 Jul 19 10:26 llvmlite
drwxr-sr-x  2 root staff   4096 Jul 19 10:26 llvmlite-0.19.0.dist-info

$ /home/amit/numba# python -m numba.runtests numba.tests.test_dispatcher.TestCache

........ipython version: 5.4.1
...FFE...
======================================================================
ERROR: test_non_writable_pycache (numba.tests.test_dispatcher.TestCache)
----------------------------------------------------------------------
OSError: [Errno 2] No such file or directory: '/tmp/numba-tests.0/test_cache-_knByx/__pycache__'

======================================================================
FAIL: test_non_creatable_pycache (numba.tests.test_dispatcher.TestCache)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "numba/tests/test_dispatcher.py", line 956, in test_non_creatable_pycache
    self._test_pycache_fallback()
  File "numba/tests/test_dispatcher.py", line 946, in _test_pycache_fallback
    self.check_pycache(0)
  File "numba/tests/test_dispatcher.py", line 644, in check_pycache
    self.assertEqual(len(c), n, c)
AssertionError: 2 != 0 : ['dispatcher_caching_test_fodder.add_usecase-27.py27.nbi', 'dispatcher_caching_test_fodder.add_usecase-27.py27.1.nbc']

======================================================================
FAIL: test_non_writable_pycache (numba.tests.test_dispatcher.TestCache)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "numba/tests/test_dispatcher.py", line 968, in test_non_writable_pycache
    self._test_pycache_fallback()
  File "numba/tests/test_dispatcher.py", line 946, in _test_pycache_fallback
    self.check_pycache(0)
  File "numba/tests/test_dispatcher.py", line 644, in check_pycache
    self.assertEqual(len(c), n, c)
AssertionError: 2 != 0 : ['dispatcher_caching_test_fodder.add_usecase-27.py27.nbi', 'dispatcher_caching_test_fodder.add_usecase-27.py27.1.nbc']

----------------------------------------------------------------------
Ran 16 tests in 5.865s

FAILED (failures=2, errors=1)
```
For the error shown above , seems like file creation is failing
```
OSError: [Errno 2] No such file or directory: '/tmp/numba-tests.0/test_cache-_knByx/__pycache__'
```
and failures are due to assertion errors
```
AssertionError: 2 != 0 : ['dispatcher_caching_test_fodder.add_usecase-27.py27.nbi', 'dispatcher_caching_test_fodder.add_usecase-27.py27.1.nbc']
```

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -1007,7 +1007,7 @@ class TestCache(BaseCacheUsecasesTest):
             cmd = ['chattr', '+i', self.tempdir]
             # redirect stderr to stdout and use check_output as a noise sink
             subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-        except subprocess.CalledProcessError:
+        except Exception:
             pass
         else:
             cmd = ['chattr', '-i', self.tempdir]
@@ -1030,7 +1030,7 @@ class TestCache(BaseCacheUsecasesTest):
             cmd = ['chattr', '+i', pycache]
             # redirect stderr to stdout and use check_output as a noise sink
             subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-        except subprocess.CalledProcessError:
+        except Exception:
             pass
         else:
             cmd = ['chattr', '-i', pycache]

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -77,9 +77,16 @@ class BaseTest(TestCase):
         return f, check
 
 def check_access_is_preventable():
+    # This exists to check whether it is possible to prevent access to
+    # a file/directory through the use of `chmod 500`. If a user has
+    # elevated rights (e.g. root) then writes are likely to be possible
+    # anyway. Tests that require functioning access prevention are
+    # therefore skipped based on the result of this check.
     tempdir = temp_directory('test_cache')
     test_dir = (os.path.join(tempdir, 'writable_test'))
     os.mkdir(test_dir)
+    # assume access prevention is not possible
+    ret = False
     # check a write is possible
     with open(os.path.join(test_dir, 'write_ok'), 'wt') as f:
         f.write('check1')
@@ -88,10 +95,14 @@ def check_access_is_preventable():
     try:
         with open(os.path.join(test_dir, 'write_forbidden'), 'wt') as f:
             f.write('check2')
-    except PermissionError:
-        ret = True
-    else:
-        ret = False
+    except (OSError, IOError) as e:
+        # Check that the cause of the exception is due to access/permission
+        # as per https://github.com/conda/conda/blob/4.5.0/conda/gateways/disk/permissions.py#L35-L37
+        eno = getattr(e, 'errno', None)
+        if eno in (errno.EACCES, errno.EPERM):
+            # errno reports access/perm fail so access prevention via
+            # `chmod 500` works for this user.
+            ret = True
     finally:
         os.chmod(test_dir, 0o775)
         shutil.rmtree(test_dir)

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -76,6 +76,31 @@ class BaseTest(TestCase):
         f = jit(**self.jit_args)(pyfunc)
         return f, check
 
+def check_access_is_preventable():
+    tempdir = temp_directory('test_cache')
+    test_dir = (os.path.join(tempdir, 'writable_test'))
+    os.mkdir(test_dir)
+    # check a write is possible
+    with open(os.path.join(test_dir, 'write_ok'), 'wt') as f:
+        f.write('check1')
+    # now forbid access
+    os.chmod(test_dir, 0o500)
+    try:
+        with open(os.path.join(test_dir, 'write_forbidden'), 'wt') as f:
+            f.write('check2')
+    except PermissionError:
+        ret = True
+    else:
+        ret = False
+    finally:
+        os.chmod(test_dir, 0o775)
+        shutil.rmtree(test_dir)
+    return ret
+
+_access_preventable = check_access_is_preventable()
+_access_msg = "Cannot create a directory to which writes are preventable"
+skip_bad_access = unittest.skipUnless(_access_preventable, _access_msg)
+
 
 class TestDispatcher(BaseTest):
 
@@ -994,6 +1019,7 @@ class TestCache(BaseCacheUsecasesTest):
         # wouldn't be met)
         self.check_pycache(0)
 
+    @skip_bad_access
     @unittest.skipIf(os.name == "nt",
                      "cannot easily make a directory read-only on Windows")
     def test_non_creatable_pycache(self):
@@ -1002,19 +1028,9 @@ class TestCache(BaseCacheUsecasesTest):
         os.chmod(self.tempdir, 0o500)
         self.addCleanup(os.chmod, self.tempdir, old_perms)
 
-        # if run as root dir perms have no effect, so call chattr +i
-        try: # will fail if not root
-            cmd = ['chattr', '+i', self.tempdir]
-            # redirect stderr to stdout and use check_output as a noise sink
-            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-        except Exception:
-            pass
-        else:
-            cmd = ['chattr', '-i', self.tempdir]
-            self.addCleanup(subprocess.check_output, cmd, stderr=subprocess.STDOUT)
-
         self._test_pycache_fallback()
 
+    @skip_bad_access
     @unittest.skipIf(os.name == "nt",
                      "cannot easily make a directory read-only on Windows")
     def test_non_writable_pycache(self):
@@ -1024,17 +1040,6 @@ class TestCache(BaseCacheUsecasesTest):
         old_perms = os.stat(pycache).st_mode
         os.chmod(pycache, 0o500)
         self.addCleanup(os.chmod, pycache, old_perms)
-
-        # if run as root dir perms have no effect, so call chattr +i
-        try: # will fail if not root
-            cmd = ['chattr', '+i', pycache]
-            # redirect stderr to stdout and use check_output as a noise sink
-            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
-        except Exception:
-            pass
-        else:
-            cmd = ['chattr', '-i', pycache]
-            self.addCleanup(subprocess.check_output, cmd, stderr=subprocess.STDOUT)
 
         self._test_pycache_fallback()
 

--- a/numba/tests/test_dispatcher.py
+++ b/numba/tests/test_dispatcher.py
@@ -1002,6 +1002,17 @@ class TestCache(BaseCacheUsecasesTest):
         os.chmod(self.tempdir, 0o500)
         self.addCleanup(os.chmod, self.tempdir, old_perms)
 
+        # if run as root dir perms have no effect, so call chattr +i
+        try: # will fail if not root
+            cmd = ['chattr', '+i', self.tempdir]
+            # redirect stderr to stdout and use check_output as a noise sink
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            pass
+        else:
+            cmd = ['chattr', '-i', self.tempdir]
+            self.addCleanup(subprocess.check_output, cmd, stderr=subprocess.STDOUT)
+
         self._test_pycache_fallback()
 
     @unittest.skipIf(os.name == "nt",
@@ -1013,6 +1024,17 @@ class TestCache(BaseCacheUsecasesTest):
         old_perms = os.stat(pycache).st_mode
         os.chmod(pycache, 0o500)
         self.addCleanup(os.chmod, pycache, old_perms)
+
+        # if run as root dir perms have no effect, so call chattr +i
+        try: # will fail if not root
+            cmd = ['chattr', '+i', pycache]
+            # redirect stderr to stdout and use check_output as a noise sink
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            pass
+        else:
+            cmd = ['chattr', '-i', pycache]
+            self.addCleanup(subprocess.check_output, cmd, stderr=subprocess.STDOUT)
 
         self._test_pycache_fallback()
 

